README.md
# scp-cf-database


This class simplefies the use of working with the @sap/hana-client
- The connection will be created automatically, no need to set it up manually.
- All queries will be escaped automatically
- See the examples below and checkout the example/example-testrun.js file


### Require in JS
```js
const DB = require("scp-cf-database");
```


### query
Exec a query
```js
const a = await DB.query(`SELECT * FROM "table.Example" WHERE "foo" = ?`, ["bar"]);
```

### batch query 
Exec a batch query
```js
const i = await DB.queryBatch('INSERT INTO "table.Example"("foo", "example") VALUES (?, ?)', [['bar', 'text'], ['bar 2', 'text 2']]);
console.log(i); // 2 // no affected rows 
```

### insert
Inserting will return the created insert
```js
const a = await DB.insert("table.Example", { foo: "bar"});
console.log(a); // { id: 1234, foo: "bar"}
```

### batch insert
A batch insert will return the no of affected rows
```js
const i = await DB.insertBatch(
	"table.Example", 
	[
		{ "foo": "bar", "example": "text" },
		{ "foo": "bar 2", "example": "text 2" }
	]
);
console.log(i); // 2 // no affected rows 
```

### update
UPDATE "table.Example" set "foo" = 'bar', "example" = 'text' WHERE "id" = 1
```js
await DB.update(
	"table.Example", //tabel
	{ foo: "bar", example: "text"}, // data
	`"id" = ?`, //filter
	[1] // filter waarden
);
```

### delete
DELETE FROM "table.Example" WHERE "id" = '1'
```js
await DB.delete("table.Example",	`"id" = ?`, [1]);
```

### select
```js
await DB.select("table.Example", "*", `"foo" = ?`, ["bar"]);
```

### callProcedure
```js
const aResult = await DB.callProcedure(
	"createPerson", // Procedure naam
	{ IM_FIRSTNAME: "Barry", IM_LASTNAME: "Dam" }, // Importing paams
	["EX_FULLNAME"] // Exporting params
);
console.log(aResult.EX_FULLNAME);
```

### callFunction
```js
await DB.callFunction("functionname", mImporting);
```


CHANGELOG.md
# 1.1.4 - 1.1.5
###### 03-05-2021
Op het moment dat een query fout gaat kan nu in de catch bij .sqlQuery en .sqlParams de query zien waarbij het fout is gegaan 

# 1.1.1 - 1.1.3
###### 30-03-2021
sap hana-client package update 2.7.26

# 1.1.0
###### 17-11-2020
[FEATURE] nieuwe methode queryBatch en static method insertBatch toegevoegd waarmee batches kunnen worden gedaan.

# 1.0.5
###### 13-11-2020
[FIX] Er kon zich een probleem voordoen met het geheugen van de DB.
Je kreeg dan een melding '{"error":"cannot allocate enough memory: Out of memory on query open"}'
Daarom his de Connection exec functie herschreven zodat het statement altijd gedropped wordt en dus het geheugen gecleared.

# 1.0.4
###### 14-07-2020
@sap/hana-client versie bijgewerkt naar de laatste versie 2.5.101

# 1.0.2-3
###### 29-06-2020
[FIX] Procedures aanroepen werkte niet, dat is nu verholpen
@sap/hana-client versie bijgewerkt naar de laatste versie 2.5.86
@sap/xsenv versie bijgewerkt naar de laatste versie 3.0.0"

# 1.0.1
###### 31-03-2020
@sap/hana-client versie bijgewerkt naar de laatste versie 2.4.182

# 1.0.0
###### 08-01-2020
Eerste versie beschikbaar als node package

example/example-testrun.js
"use strict";
const SapCfDb			= require("../modules/sap-cf-database"),
	{serializeError }	= require('serialize-error'),
	{ format }			= require("util");
module.exports = function(oApp){ 

	const wrap = fn => (...args) => fn(...args).catch(args[2]);
	
	oApp.get("/", (oReq, oRes) => {
		oRes.send("go to /testall");
	});
	
	
	
	const fnAddDBLOG = (oReq, sAction, vResult) => {
		if (!("_DBLOG" in oReq)) { oReq._DBLOG = {} }
		oReq._DBLOG[sAction] = vResult;
	}
	
	oApp.get(
		"/testall", 
		wrap(async (oReq, oRes, fnNext) => { // insert
			oReq._startTime = new Date();
			const m = await SapCfDb.insert("sapCfDatabase.address", {
				street			: "Dijkgraaf",  
				no 				: 4,
				postalCode		: "6921RL",
				city			: "Duiven"
			});
			fnAddDBLOG(oReq, "insert", m);
			fnNext();
		}),
		wrap(async (oReq, oRes, fnNext) => { // select
			const a1 = await SapCfDb.select(
				"sapCfDatabase.address",
				"*",
				'"id" = ?',
				[oReq._DBLOG.insert.id]
			);
			const a2 = await SapCfDb.select(
				"sapCfDatabase.address",
				'"street", "no", "city"',
				'"id" = ?',
				[oReq._DBLOG.insert.id]
			);
			const a3 = await SapCfDb.select(
				"sapCfDatabase.address",
				{ id: oReq._DBLOG.insert.id }
			);
			fnAddDBLOG(oReq, "read", {
				"example1": a1,
				"example2": a2,
				"example3": a3
			});
			fnNext();
		}),
		wrap(async (oReq, oRes, fnNext) => { // update
			const v1 = await SapCfDb.update(
				"sapCfDatabase.address",
				{ street: "Fotograaf" },
				'"id" = ?',
				[oReq._DBLOG.insert.id]
			);
			const v2 = await SapCfDb.update(
				"sapCfDatabase.address",
				{ street: "Fotograaf" },
				{ id: oReq._DBLOG.insert.id }
			);
			const v3 = await SapCfDb.select(
				"sapCfDatabase.address",
				{ id: oReq._DBLOG.insert.id }
			);
			fnAddDBLOG(oReq, "update", {
				"example1": v1,
				"example2": v2,
				"example3": v3
			});
			fnNext();
		}),
		wrap(async (oReq, oRes, fnNext) => { // delete
			const vResult = await SapCfDb.delete(
				"sapCfDatabase.address",
				{ id: oReq._DBLOG.insert.id }
			);
			fnAddDBLOG(oReq, "delete", vResult);
			fnNext();
		}),
		(oReq, oRes) => {
			fnAddDBLOG(oReq, "executionTime", format('%dms', new Date() - oReq._startTime));
			oRes.json(oReq._DBLOG);
		}
	);
	
	oApp.get("/testInsert", wrap(async (oReq, oRes) => {
		const aInserts = [
			{
				street			: "Dijkgraaf",  
				no 				: 4,
				postalCode		: "6921RL",
				city			: "Duiven"
			},
			{
				street			: "Dijkgraaf",  
				no 				: 123,
				postalCode		: "6921RL",
				city			: "Duiven"
			}
		];
		const aResult = await Promise.all(aInserts.map(m => SapCfDb.insert("sapCfDatabase.address", m)));
		oRes.json(aResult);
	}));
	
	oApp.get("/testInsertBatch", wrap(async (oReq, oRes) => {
		const result = await SapCfDb.queryBatch(
			'INSERT INTO "sapCfDatabase.address"("street", "no", "postalCode", "city") VALUES (?, ? ,? ,?)', 
			[
				["Frans Arnoldstraat", 13, "6433DH", "Hoensbroek"],
				["Leemkampsweg", 5, "6415RP", "Heerlen"],
			]
		);
		const dbentries = await SapCfDb.select("sapCfDatabase.address");
		const deleted = await SapCfDb.delete("sapCfDatabase.address");
		oRes.json({
			result,
			dbentries,
			deleted
		});
	}));
	
	oApp.get("/testInsertBatch2", wrap(async (oReq, oRes) => {
		const aInserts = [
			{
				street			: "Dijkgraaf",  
				no 				: 4,
				postalCode		: "6921RL",
				city			: "Duiven"
			},
			{
				street			: "Dijkgraaf",  
				no 				: 123,
				postalCode		: "6921RL",
				city			: "Duiven"
			}
		];
		const result = await SapCfDb.insertBatch("sapCfDatabase.address", aInserts);
		const dbentries = await SapCfDb.select("sapCfDatabase.address");
		const deleted = await SapCfDb.delete("sapCfDatabase.address");
		oRes.json({
			result,
			dbentries,
			deleted
		});
	}));
	
	/**
	 * /read
	 */
	oApp.get("/testf", wrap(async (oReq, oRes) => {
		const a = await SapCfDb.query('SELECT * from "sapCfDatabase.address"');
		oRes.json(a);
	}));
	
	/**
	 * Procedure
	 */
	 oApp.get("/test-procedure", wrap(async(oReq, oRes) => {
	 	const v = await SapCfDb.callProcedure("PROCEDURETEST", { ID_TEST : 1 }, ["ID_OUT","EXAMPLE"]);
	 	oRes.json(v);
	 }));
	
	/**
	 * Error handling
	 */
	oApp.use((oErr, oReq, oRes, fnNext) => {
		oRes.status(400).json(serializeError(oErr));
	});
	
	return  oApp;
	
	


	
	
};



lib/index.js
/* eslint-env es8 */
/* eslint no-console: 0*/
"use strict";
/****************
 * Database class die eenvoudig in alle andere node modules gebruikt kan worden.
 * - De connectie wordt automatisch gelegd in de constructor dus daar hoef je niet meer naar te kijken
 * - Alle queries worden automatisch escaped
 * 
 * @class scp-cf-database 
 * @requires @sap/xsenv
 * @requires @sap/hana-client
 * @author Barry Dam
 * @version	1.1.4
 */
const xsenv					= require('@sap/xsenv'),
	hanaClient				= require('@sap/hana-client'),
	{ createProcStatement } = require('@sap/hana-client/extension/Stream'),
	{ promisify }			= require("util");
	
const _private = {
	weakmap : new WeakMap(),// gebruik een WeakMap voor private data en het voorkomen van memoryleaks	
	/**
 	 * Converteren van array parameters naar statement ? ? ? 
 	 * @param {array} aParams - Query parameters (voorbeeld: ["id", "naam"])
 	 * @return {string} - voorbeeld: "?, ?"
 	 */
	prepareParamsForStatement  : (aParams) => {
		const iParamLength	= aParams.length;
		let	sParams 		= "";
		// converteren van parameters
		if (aParams && iParamLength > 0) {
			for (const i in aParams)  {
				sParams += (i !== (iParamLength-1).toString()) ? "?, ": "?";
			}
		}
		return sParams;
	},
	staticDB		: null
};

class SapCfDatabase {
	/**
 	 * Class constructor
 	 */
	constructor(mParams) {
		if (typeof mParams !== "object") { mParams = {}; }
		const mEnvOptions = xsenv.getServices({ hana: { tag: "hana" } });
		_private.weakmap.set(this, {
			connOptions: {
				serverNode			: `${mParams.host || mEnvOptions.hana.host}:${mParams.port || mEnvOptions.hana.port}`,
				uid					: mParams.user || mEnvOptions.hana.user,
				pwd					: mParams.password || mEnvOptions.hana.password,
				encrypt				: ("encrypt" in mParams) ? mParams.encrypt : true,
				sslCryptoProvider	: mParams.sslCryptoProvider || "openssl",
				sslTrustStore		: mParams.SSLCertificate || `${mEnvOptions.hana.certificate}`,
				currentschema		: mParams.schema || mEnvOptions.hana.schema
	       }
		});
	}
	
	/**
	 * Connectie object (hanaClient connection)
	 * De connect functie connecteerd naar de connectie opties die zijn opgegeven bij het aanroepen van de class
	 * De volgende functies worden naar een promise omgezet:
	 * - connect
	 * - exec
	 * - disconnect
	 */
	get Connection() {
		const mWM = _private.weakmap.get(this);
		if (!("connection" in mWM)) {
			const oConnection = hanaClient.createConnection();
			if (!oConnection) { throw "Fout bij het opzetten van de connectie van de Database"; }
			// connect omzetten naar promise functie en automatisch connecteren met de ingestelde connOptions
			oConnection.connectionStart = (fnOrig => {
				return mConnectOptions => fnOrig(mConnectOptions || mWM.connOptions);
			})(oConnection.connect.bind(oConnection));
			// onderstaande functies allemaal omzetten naar een promise
			const fnPrepare			= promisify(oConnection.prepare.bind(oConnection));
			oConnection.connectionDisconnect		= promisify(oConnection.disconnect.bind(oConnection));
			// exec omschrijven zodat statement gebruikt wordt en drop er in zit
			oConnection.connectionExec = async (sQuery, aParams) => { //eslint-disable-line
				let oStmt;
				try {
					oStmt = await fnPrepare(sQuery);
				} catch (err) {
					err.sqlQuery 	= sQuery;
					err.sqlParams 	= aParams;
					throw err;
				}
				return new Promise((fnResolve, fnReject) => {
					oStmt.exec(aParams, (err, mResult) => {
						if (err) { 
							err.sqlQuery 	= sQuery;
							err.sqlParams 	= aParams;
							return fnReject(err); 
						}	
						oStmt.drop(err => {
							if (err) { return fnReject(err); }		
							fnResolve(mResult);
						});
					});
				});
            }
            // exec batch omschrijven zodat statement gebruikt wordt en  drop er in zit
			oConnection.connectionExecBatch = async (sQuery, aParams) => { //eslint-disable-line
				let oStmt;
				try {
					oStmt = await fnPrepare(sQuery);
				} catch (err) {
					err.sqlQuery 	= sQuery;
					err.sqlParams 	= aParams;
					throw err;
				}
				return new Promise((fnResolve, fnReject) => {
					oStmt.execBatch(aParams, (err, mResult) => {
						if (err) { 
							err.sqlQuery 	= sQuery;
							err.sqlParams 	= aParams;
							return fnReject(err); 
						}
						oStmt.drop(err => {
							if (err) { return fnReject(err); }		
							fnResolve(mResult);
						});
					});
				});
            }
			// procedures aanroepen
			oConnection.callProcedure	= (sQuery, mIn) => {
				return new Promise((fnResolve, fnReject) => {
					createProcStatement(oConnection, sQuery, function(err, oStmt) {
						if (err) { return fnReject(oStmt); }
						oStmt.exec(mIn, function(err, mResult) {
							if (err) { return fnReject(oStmt); }
							oStmt.drop(err => {
								if (err) { return fnReject(oStmt); }		
								fnResolve(mResult);
							});
						});
					});
				});
			}
			// resave weakmap
			mWM.connection = oConnection;
			_private.weakmap.set(this, mWM);
		}
		return mWM.connection;
	}
	
	/**
	 * @param {bool} [mOptions.autoDisconnect=true] 
	 */
	async execQuery(sQuery, aParameters, mOptions) { //es-lint-disable-line
		const mOpt	= { ...{autoDisconnect:true}, ...((typeof mOptions === "object") ? mOptions : {}) };
		if (this.Connection.state() === "disconnected") {
			await this.Connection.connectionStart();
		}
		let vResult;
		if (sQuery.toUpperCase().startsWith("CALL") && ! Array.isArray(aParameters)) {
			vResult = await this.Connection.callProcedure(sQuery, aParameters);
		} else {
			let	aParams = aParameters || [];
			if (! Array.isArray(aParams)) { aParams = [aParams]; }
			vResult = await this.Connection.connectionExec(sQuery, aParams);
		}
		if (mOpt.autoDisconnect) { this.Connection.connectionDisconnect(); }
		return vResult;
    }	

    /**
	 * Uitvoeren van een batch opdracht
	 */
	async execBatch(sQuery, aData, { autoDisconnect = true} = {}) { //es-lint-disable-line
		if (this.Connection.state() === "disconnected") {
			await this.Connection.connectionStart();
		}
		let	aValues = aData || [];
        if (! Array.isArray(aValues)) { aValues = [aValues]; }
        const vResult = await this.Connection.connectionExecBatch(sQuery, aValues);
        if (autoDisconnect) { this.Connection.connectionDisconnect(); }        
		return vResult;
    }
	
	/**
	 * Query de connectie wordt automatisch gesloten na 1/2 minuut van inactiviteit
	 */
	static async query(sStatement, aParameters, { queryBatch = false } = {}) {
		if (! _private.staticDB) { 
			_private.staticDB					= new SapCfDatabase();
			_private.staticDB.runIds			= [];
			_private.staticDB.timeoutRemoveDb	= null;
			_private.staticDB.onRunDone
			const fnOnDone = sRunId => {
				_private.staticDB.runIds = _private.staticDB.runIds.filter(s => s !== sRunId); 
				if (_private.staticDB.runIds.length === 0) {
					if (_private.staticDB.timeoutRemoveDb) { clearTimeout(_private.staticDB.timeoutRemoveDb); }
					_private.staticDB.timeoutRemoveDb = setTimeout(_ => {
						_private.staticDB.Connection.connectionDisconnect();
						_private.staticDB = null;
					}, 30000); // auto kill connection na 1/2 minuut inactiviteit
				}
			};
			_private.staticDB.startRun			= async (sStatement, aParameters, { autoDisconnect = false, queryBatch = false } = {}) => {
				if (_private.staticDB.timeoutRemoveDb) { clearTimeout(_private.staticDB.timeoutRemoveDb); }
				const sRunId = `_${Math.random().toString(36).substr(2, 9)}`;
				_private.staticDB.runIds.push(sRunId);
				try {
					const fnQueryMethod = (queryBatch) ? _private.staticDB.execBatch : _private.staticDB.execQuery ;
					const vResult = await fnQueryMethod.apply(_private.staticDB, [sStatement, aParameters, { autoDisconnect }]);  //eslint-disable-line
					fnOnDone(sRunId);
					return vResult;
				} catch(e) {
					fnOnDone(sRunId);
					throw e;
				}
			}
		}
		return _private.staticDB.startRun(sStatement, aParameters, { autoDisconnect: false, queryBatch }); 
	}
	
	static queryBatch(sStatement, aParameters) {
		return this.query(sStatement, aParameters, { queryBatch : true })
	}
	
	/**
	 * Function procedure aanroepen
	 * @param {string} sName - Functienaam
	 * @param {array} aParams - Functie import variabelen
	 * @return {Promise} 
	 */
	static callFunction(sName, aParams) {
		aParams = aParams || [];
		return this.query(
			`SELECT * FROM "${sName}"(${_private.prepareParamsForStatement(aParams)})`,
			aParams
		);
	}
	
	/**
	 * Prodecure aanroepen
	 * @param {string} sName					- Functienaam
	 * @param {array} mData						- Functie import variabelen
	 * @param {string|array) vExportingParams	- Exporting variablen
	 * @return {Promise} - met het resultaat zie hieronder
	 * @example waar vExportingParams een string is
	 * await callProcedure("examplePerson", { IM_FIRSTNAME: "Barry", IM_LASTNAME: "Dam" }, "EX_FULLNAME"); //result == Barry Dam } 
	 * @example waar vExportingParams een Array is > het result zal een object zijn waarbij de keys de Exporting parameters zijn
	 * await callProcedure("examplePerson", { IM_FIRSTNAME: "Barry", IM_LASTNAME: "Dam" }, ["EX_FULLNAME", "EX_AGE"])  //result == {EX_FULLNAME: "Barry Dam", EX_AGE: "31"}
	 */
	static async callProcedure(sName, mData, vExportingParams) {
		if (typeof mData !== "object") { throw "Check Funtcion importing params object not "; }
		const
			aColumns			= Object.keys(mData),
			aExportingParams	= (typeof vExportingParams === "string") ? [vExportingParams] : vExportingParams,
			sColumns			= (aColumns.length) ? `${aColumns.join(' => ?, ')} => ?` : "",
			sExporting			= (sColumns && Array.isArray(aExportingParams) && aExportingParams.length) ? " ," + aExportingParams.join(" => ?, ") + " => ?" : "", //eslint-disable-line
			mResult 			= await this.query(`CALL "${sName}"(${sColumns} ${sExporting})`, mData);
		if (mResult && typeof vExportingParams === "string") { 
			return mResult[Object.keys(mResult).find(key => key.toLowerCase() === vExportingParams.toLowerCase())];
		} else if (mResult && Array.isArray(aExportingParams) && aExportingParams.length) {
			return aExportingParams.reduce((a, b) => (a[b] = mResult[Object.keys(mResult).find(key => key.toLowerCase() === b.toLowerCase())]), {});
		} 
		return mResult;
	}
	
	/**
	 * Insert nieuwe gegevens in de DB
	 * @param {string} sTablename - Tabel naam *schema name niet verplicht
	 * @param {map} mData - data dat wordt toegevoegd aan de db { kolomNaam: "waarde", kolomNaam2: "Waarde 2" }
	 * @return {Promise} die het zojuist aangemaakte resultaat retourneert 
	 * Voorbeeld normaal
	 * await db.insert("Example", { foo : "bar" }) // { id: 123, foo: "bar"}  });
	 * resultaat { id: 123, foo : "bar" };
	 * Voorbeeld batch insert
	 * await db.insert("Example", [{ foo : "bar" }, { foo : "bar 2" }]);
	 * resultaat [{ id: 123, foo : "bar" }, { id: 124, foo : "bar 2" }];
	 */
	static insert(sTablename, vData) {
		if (Array.isArray(vData)) { // batchmode
			return Promise.all(vData.map(mData => _insertqueue.addInsert(sTablename, mData)));  //eslint-disable-line
		}
		return _insertqueue.addInsert(sTablename, vData); //eslint-disable-line
	}
	
	/**
	 * Insert een batch met nieuwe gegevens in de DB
	 * @param {string} sTablename - Tabel naam *schema name niet verplicht
	 * @param {array} array met dataa - data dat wordt toegevoegd aan de db [{ kolomNaam: "waarde", kolomNaam2: "Waarde 2" }, { kolomNaam: "waarde", kolomNaam2: "Waarde 2" }]
	 * @return {integer} affected rows
	 */
	static async insertBatch(sTableName, aBatchData) {
		if (!aBatchData || !aBatchData.length || typeof aBatchData[0] !== "object") {
			throw new Error("DB insertBatch error: Invalid batch data");
		}
		const aValues = [], 
			sColumns = Object.keys(aBatchData[0]).join(",");
		// batchdata controleren
		for(const i in aBatchData) { // for loop for performance
			aValues.push(Object.values(aBatchData[i]));		
			if (sColumns !== Object.keys(aBatchData[i]).join(",")) {
				throw new Error("DB insertBatch error: All batchdata should contain the same columns in the same order");
			}
		}
		return await this.queryBatch(
			`INSERT INTO "${sTableName}"("${Object.keys(aBatchData[0]).join('", "')}") VALUES (${_private.prepareParamsForStatement(Object.keys(aBatchData[0]))})`, 
			aValues
		);
	}
	
	/**
	 * Select
	 * @param {string}		sTablename				- Tabel naam *schame name niet verplicht
	 * @param {string}		sColumns=*				- Kolommen die geselecteerd dienen te worden
	 * @param {string}		sWhereANdOrderEtc		- Where string. note: indien je geen where nodig hebt maar wel een order doe dan '1=1 ORDERBY "Test"'
	 * @param {array}		[aEscapeWhereValues]	- ? Waarden voor de where string die moeten worden escaped
	 * @param {fnQueryCallback}	fnCallback				- callback functie 2e parameter geeft een array met resultaten terug
	 *
	 * Voorbeeld 1:
	 * database.select("Example", "*", '"id" = ? AND "foo" = \'bar\'', [1], function(error, aData) { // callback });
	 * 
	 * voorbeeld 2
	 * database.select("Example", "*", { id = 1 }, function(error, aData) { // callback });
	 * 
	 * voorbeeld 3
	 * database.select("Example", { id : 1 });
	 * 
	 * voorbeeld 3a 
	 * database.select("Example", { id : 1, foo: "bar" }); // and 
	 *
	 * voorbeeld 3b
	 * database.select("Example", { id : 1, foo: "bar" }, false); // or 
	 */
	static select(sTablename, sColumns, sWhereAndOrderEtc, aEscapeWhereValues) {
		let aValues 	= [];
		if (typeof sColumns === "object") { // voorbeeld 3
			sWhereAndOrderEtc = sColumns;
			sColumns = "*";
		}
		if (typeof sWhereAndOrderEtc === "object") { // voorbeeld 2
			aValues 			= Object.values(sWhereAndOrderEtc);
			sWhereAndOrderEtc	= `"${Object.keys(sWhereAndOrderEtc).join(`" = ? ${((!aEscapeWhereValues) ? "and" : "or")} "`)}" = ?`;
		} else if (Array.isArray(aEscapeWhereValues) && aEscapeWhereValues.length) { // voorbeeld 1
			aValues = aEscapeWhereValues;
		}
		return this.query(
			`SELECT ${sColumns || "*"} FROM "${sTablename}"${(sWhereAndOrderEtc) ?  ` WHERE ${sWhereAndOrderEtc}` : ''}`,
			aValues
		);
	}
	
	/**
	 * Update de DB
	 * @param {string}		sTablename				- Tabel naam *schema name niet verplicht
	 * @param {map} 		mData					- Data dat moet worden bijgewerkt { kolomNaam: "waarde", kolomNaam2: "Waarde 2" }
	 * @param {string}		[sWhere=""]				- Where string "id" = ? AND "foo" = 'bar' // met ? of zonder kan beide
	 * @param {array}		[aEscapeWhereValues]	- ? Waarden voor de where string die moeten worden escaped
	 * 
	 * update(sTablename, mData, sWhere, fnCallback)
	 * Kan ook zonder sWhere en aEscapeWhereValues, met callback:
	 * update(sTablename, mData, fnCallback)
	 * 
	 * Voorbeeld 1:
	 * database.update("Example", { foo : "bar" }, '"id" = ? AND "foo" = \'bars\'', [1]);
	 * 
	 * voorbeeld 2
	 * database.update("Example", { foo : "bar" }, { id = 1 });
	 */
	static update(sTablename, mData, sWhere, aEscapeWhereValues) {
		const	aColumns	= Object.keys(mData);
		let		aValues			= Object.values(mData);
		if (typeof sWhere === "object") { // Voorbeeld 2
			aValues = [...aValues, ...Object.values(sWhere)];
			sWhere	= `"${Object.keys(sWhere).join("\" = ? and \"")}" = ?`;
		} else if (Array.isArray(aEscapeWhereValues) && aEscapeWhereValues.length) { // Voorbeeld 1
			aValues = [...aValues, ...aEscapeWhereValues];
		}
		return this.query(
			`UPDATE "${sTablename}" SET "${aColumns.join('" = ?, "')}" = ? ${((sWhere) ?  ` WHERE ${sWhere}` : '' )}`,
			aValues
		);
	}
	
	/**
	* Verwijderen van gegevens
	* @param {string} 	sTablename				- Tabel naam *schema name niet verplicht
	* @param {sWhere} 	sWhere					- Where string "id" = ? AND "foo" = 'bar' // met ? of zonder kan beide
	* @param {array}		[aEscapeWhereValues]	- ? Waarden voor de where string die moeten worden escaped
	* 
	* Kan ook zonder aEscapeWhereValues
	* delete(sTablename, sWhere)
	* 
	* Voorbeeld:
	* delete("Example", '"id" = ?', [1]);
	* Voorbeeld 2:
	* delete("Example", {id : 1});
	*/
	static delete(sTablename, sWhere, aEscapeWhereValues) {
		let aValues	 = [];
		if (typeof sWhere === "object") {
			aValues = Object.values(sWhere);
			sWhere	= `"${Object.keys(sWhere).join("\" = ? and \"")}" = ?`;
		} else if (Array.isArray(aEscapeWhereValues) && aEscapeWhereValues.length) { // where velden toevoegen (if any)
			aValues = aEscapeWhereValues;
		}
		return this.query(
			`DELETE FROM "${sTablename}"` + ((sWhere) ?  " WHERE " + sWhere : "" ), // eslint-disable-line 
			aValues
		);
	}
}

/**
 * De Database.insert functie geeft als resultaat de db rij terug (met een eventuele primarykey waarde)
 * Om ervoor te zorgen dat de primarykey waarde klopt, moeten deze inserts synchroon uitgevoerd worden
 * de insertqueue zorgt dat dat gebeurd.
 */
const _insertqueue	 = {
	_queue		: {
		/*
			"tablename" : [
				{ data : {foo: "bar"}, addCallback: function },
				...
			]
		*/
	},
	_running	: [], 
	addInsert: (sTableName, mData) => {
		if (!(sTableName in _insertqueue._queue)) {
			_insertqueue._queue[sTableName] = [];
		}
		const oPromise = new Promise((fnResolve, fnReject) => {
			_insertqueue._queue[sTableName].push({
				data	: mData,
				promise	: {
					resolve	: fnResolve,
					reject	: fnReject 
				}
			});
		});
		_insertqueue.runNext(sTableName);
		return oPromise;
	},
	runNext: async (sTableName) => {
		if (_insertqueue._running.indexOf(sTableName) !== -1) { return; }
		const mNext = _insertqueue._queue[sTableName].shift();
		if (!mNext) { return; }
		_insertqueue._running.push(sTableName);
		try {
			const mResult = await _insertqueue.execute(sTableName, mNext.data);
			mNext.promise.resolve(mResult);
		} catch(o) { mNext.promise.reject(o); }
		_insertqueue._running = _insertqueue._running.filter(s => s !== sTableName);
		_insertqueue.runNext(sTableName);
	},
	execute: async (sTablename, mData) => {
		const aRows 	= await SapCfDatabase.query(`SELECT "COLUMN_NAME" FROM "SYS"."TABLE_COLUMNS" WHERE "TABLE_NAME" = '${sTablename}' AND "IS_NULLABLE" = 'FALSE' AND "INDEX_TYPE" = 'FULL'`),
			sPrimaryKey = (aRows && aRows.length) ? aRows[0].COLUMN_NAME : null,
			aColumns	= Object.keys(mData),
			aValues		= Object.values(mData);
		if (sPrimaryKey && sPrimaryKey in mData && mData[sPrimaryKey] === null) {
			delete mData[sPrimaryKey];
		}
		const vResult = await SapCfDatabase.query(`INSERT INTO "${sTablename}"("${aColumns.join('", "')}") VALUES (${_private.prepareParamsForStatement(aValues)})`, aValues);
		if (!sPrimaryKey) { return vResult; }
		const aResult = await SapCfDatabase.query(`SELECT * FROM "${sTablename}" ORDER BY "${sPrimaryKey}" DESC LIMIT 1`);
		return aResult ? aResult[0] : null;
	}
};

module.exports = SapCfDatabase;